lib/resolve-package-path.js

```diff
@@ -1,12 +1,115 @@
+ * Resolve the real path for a file. Return null if does not
+ * exist or is not a file or FIFO, return the real path otherwise.
+ *
+ * Cache the result in the passed-in cache for performance, 
+ * keyed on the filePath passed in.
+ *
+ * @private
+ * @method _getRealFilePath
+ * @param cache the Cache object to cache the real path in, keyed
+ * by filePath.
+ * @param {String} filePath the (not necessarily resolved) 
+ * path to the file of interest
+ * @return {String} real path or null
+ */
+function _getRealFilePath(cache, filePath) {
+  var realPath = cache.get(filePath);
+
+  if (realPath) {
+    return realPath;
+  }
+
+  try {
+    var stat = fs.statSync(filePath);
+
+    if (stat.isFile() || stat.isFIFO()) {
+      realPath = fs.realpathSync(filePath);
+    } else {
+      realPath = null; 
+    }
+  } catch (e) {
+    if (
+      e !== null &&
+      typeof e === 'object' &&
+      (e.code === 'ENOENT' || e.code === 'ENOTDIR')
+    ) {
+      realPath = null; // NO PATH FOUND
+    } else {
+      throw e;
+    }
+  }
+
+  cache.set(filePath, realPath);
+
+  return realPath;
+}
+
+/*
+ * Given a package 'name' and starting directory, resolve to a real (existing) file path.
+ *
+ * Do it similar to how it is done in resolve.sync() - travel up the directory hierarchy,
+ * attaching 'node-modules' to each directory and seeing if the directory exists and
+ * has the relevant 'package.json' file we're searching for. It is *much* faster than
+ * resolve.sync(), because we don't test that the requested name is a directory.
+ * This routine assumes that it is only called when we don't already have
+ * the cached entry.
+ *
+ * @param realFilePathCache the cache containing the real paths corresponding to
+ * various file paths (which may or may not be already resolved). This is only used
+ * if passed through to _getRealFilePath.
+ * @param name the 'name' of the module, i.e. x in require(x), but with
+ *   '/package.json' on the end. It is NOT a directory (so we don't have to do the
+ *   directory checks that resolve.sync does).
+ *   NOTE: because this is an internal routine, for speed it does not check 
+ *   that '/package.json' is actually the end of the name.
+ * @param startPath the directory (non-empty, and valid) to start from, appending the name to the
+ *   directory and checking that the file exists. Go up the directory hierarchy from there.
+ *
+ * @result the path to the actual package.json file that's found, or null if not.
+ */
+function _findPackagePath(realFilePathCache, name, dir) {
+
+  var h = heimdall.start('hashForDep:_findPackagePath');
+
+  try {
+    var fsRoot = pathRoot(dir);
+
+    var currPath = dir;
+
+    while (currPath !== fsRoot) {
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
+
+      var nodeModulesPath = endsWithNodeModules ? currPath : currPath + path.sep + 'node_modules';
+      var filePath = path.join(nodeModulesPath, name);
```

Comment:
Question, since we are already bypassing `path.join` in the line above, should we here as all?

`var filePath  = nodeModulesPath + path.sep + name`

Or should we use `path.join` in both cases.

Reply:
Though it's slower (being a function), I think I favor path.join() in both cases. Thoughts?

Reply:
Ended up going with path.join in both.